x509util/loginfo.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package x509util
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/client"
+	"github.com/google/certificate-transparency-go/dnsclient"
+	"github.com/google/certificate-transparency-go/jsonclient"
+	"github.com/google/certificate-transparency-go/loglist"
+	"github.com/google/certificate-transparency-go/merkletree"
+	"github.com/google/certificate-transparency-go/tls"
+	"github.com/google/certificate-transparency-go/x509"
+)
+
+// LogInfo holds the objects needed to perform per-log verification and
+// validation of SCTs.
+type LogInfo struct {
+	Description string
+	Client      client.CheckLogClient
+	MMD         time.Duration
+	Verifier    *ct.SignatureVerifier
+	PublicKey   []byte
+
+	mu      sync.RWMutex
+	lastSTH *ct.SignedTreeHead
+}
+
+// NewLogInfo builds a LogInfo object based on a log list entry.
+func NewLogInfo(log *loglist.Log, hc *http.Client) (*LogInfo, error) {
+	// TODO(drysdale): what if a log is http:// only?
```

Comment:
Ah, I'd not spotted that part of the spec.  Fixed, and I added some protection in case the log list every becomes sane and starts putting a URL in the `url` field :-)